README.md
# tinysementer
Haskell implementation of tinysegmenter, a compact Japanese tokenizer

Machine lerning based compact Japanese tokenizer library.
Original library is written in JavaScript by Taku Kudo <http://chasen.org/~taku/software/TinySegmenter>


CHANGELOG.md
# Revision history for tinysegmenter-hs

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Text/TinySegmenter.hs
{-# LANGUAGE BangPatterns #-}
module Text.TinySegmenter where

import           Control.Monad.ST
import           Data.Bits
import           Data.Char
import qualified Data.Text                     as T
import qualified Data.Text.Array               as A
import qualified Data.Text.Internal            as TI
import qualified Data.HashMap.Strict as M
import           Data.Word

-- Markers, whose values are out of unicode code point range
b1, b2, b3, e1, e2, e3 :: Int
b1 = 0x110001
b2 = 0x110002
b3 = 0x110003
e1 = 0x110004
e2 = 0x110005
e3 = 0x110006

data Marker = U | O | B
data CTypes = TM | TH | TI | TK | TA | TN | TO

mk2i :: Marker -> Word8
mk2i m = case m of
  U -> fromIntegral $ ord 'U'
  O -> fromIntegral $ ord 'O'
  B -> fromIntegral $ ord 'B'
{-# INLINE mk2i #-}

ct2i :: CTypes -> Word8
ct2i c = case c of
  TM -> fromIntegral $ ord 'M'
  TH -> fromIntegral $ ord 'H'
  TI -> fromIntegral $ ord 'I'
  TK -> fromIntegral $ ord 'K'
  TA -> fromIntegral $ ord 'A'
  TN -> fromIntegral $ ord 'N'
  TO -> fromIntegral $ ord 'O'
{-# INLINE ct2i #-}

getCTypes :: Int -> Word8
getCTypes c
  | c `elem` m
  = ct2i TM
  | (ord '一' <= c && c <= ord '龠') || c `elem` h
  = ct2i TH
  | ord 'ぁ' <= c && c <= ord 'ん'
  = ct2i TI
  | (ord 'ァ' <= c && c <= ord 'ヴ')
    || (ord 'ｱ' <= c && c <= ord 'ﾝ') || c `elem` ksub
  = ct2i TK
  | (ord 'a' <= c && c <= ord 'z')
    || (ord 'A' <= c && c <= ord 'Z')
    || (ord 'ａ' <= c && c <= ord 'ｚ')
    || (ord 'Ａ' <= c && c <= ord 'Ｚ')
  = ct2i TA
  | (ord '0' <= c && c <= ord '9') || (ord '０' <= c && c <= ord '９')
  = ct2i TN
  | otherwise
  = ct2i TO
  where
    m    = fmap ord "一二三四五六七八九十百千万億兆"
    h    = fmap ord "々〆ヵヶ"
    ksub = fmap ord "ーｰ\xff9e"
{-# INLINABLE getCTypes #-}

takeThree :: T.Text -> (Int, Int, Int, T.Text)
takeThree text = case T.uncons text of
  Nothing -> (e1, e2, e3, text)
  Just (ca, ra) -> case T.uncons ra of
    Nothing -> (ord ca, e1, e2, ra)
    Just (cb, rb) -> case T.uncons rb of
      Nothing -> (ord ca, ord cb, e1, rb)
      Just (cc, rc) ->(ord ca, ord cb, ord cc, rc)
{-# INLINE takeThree #-}

takeOne :: T.Text -> Maybe (Int, T.Text)
takeOne = fmap (\(c, r) -> (ord c, r)) . T.uncons
{-# INLINE takeOne #-}

isPair :: Int -> Bool
isPair = (< 0x10000)
{-# INLINE isPair #-}

type Upper = Word16
type Lower = Word16
splitToWord16 :: Int -> (Upper, Lower)
splitToWord16 c = (upper, lower)
  where
    m     = c - 0x10000
    upper = fromIntegral (shiftR m 10 + 0xD800)
    lower = fromIntegral ((m .&. 0x3FF) + 0xDC00)
{-# INLINE splitToWord16 #-}

tokenToText :: [Word16] -> Int -> T.Text
tokenToText xs size = TI.text array 0 size
  where
    array = runST $ do
      arr <- A.new size
      mapM_ (\(c, i) -> A.unsafeWrite arr i c) $ zip xs [(size-1) .. 0]
      A.unsafeFreeze arr
{-# INLINABLE tokenToText #-}

data TokenizeState = TS { remain :: {-# UNPACK #-} !T.Text
                        , token :: ![Word16]
                        , tokenLength :: {-# UNPACK #-} !Int
                        , score :: {-# UNPACK #-} !Int
                        , p1 :: {-# UNPACK #-} !Word8
                        , p2 :: {-# UNPACK #-} !Word8
                        , p3 :: {-# UNPACK #-} !Word8
                        , w1 :: {-# UNPACK #-} !Int
                        , w2 :: {-# UNPACK #-} !Int
                        , w3 :: {-# UNPACK #-} !Int
                        , w4 :: {-# UNPACK #-} !Int
                        , w5 :: {-# UNPACK #-} !Int
                        , w6 :: {-# UNPACK #-} !Int
                        , c1 :: {-# UNPACK #-} !Word8
                        , c2 :: {-# UNPACK #-} !Word8
                        , c3 :: {-# UNPACK #-} !Word8
                        , c4 :: {-# UNPACK #-} !Word8
                        , c5 :: {-# UNPACK #-} !Word8
                        , c6 :: {-# UNPACK #-} !Word8
                        }

initialState :: T.Text -> TokenizeState
initialState text =
  let !(a, b, c, rmn) = takeThree text in
  TS { remain = rmn
     , token = []
     , tokenLength = 0
     , score = bias
     , p1 = mk2i U
     , p2 = mk2i U
     , p3 = mk2i U
     , w1 = b1
     , w2 = b2
     , w3 = b3
     , w4 = a
     , w5 = b
     , w6 = c
     , c1 = mk2i O
     , c2 = mk2i O
     , c3 = mk2i O
     , c4 = getCTypes a
     , c5 = getCTypes b
     , c6 = getCTypes c
     }
  where
    bias = -332
{-# INLINABLE initialState #-}



src/Text/Score.hs
module Text.Score where

import           Data.Char
import           Data.Word

b1, b2, b3, e1, e2, e3 :: Int
b1 = 0x110001
b2 = 0x110002
b3 = 0x110003
e1 = 0x110004
e2 = 0x110005
e3 = 0x110006

-- Words
h = fromIntegral $ ord 'H'
i = fromIntegral $ ord 'I'
k = fromIntegral $ ord 'K'
o = fromIntegral $ ord 'O'
a = fromIntegral $ ord 'A'
n = fromIntegral $ ord 'N'
m = fromIntegral $ ord 'M'

b = fromIntegral $ ord 'B'
u = fromIntegral $ ord 'U'

bc1 t | t == (h, h) = 6
      | t == (i, i) = 2461
      | t == (k, h) = 406
      | t == (o, h) = -1378
{-# INLINE bc1 #-}

bc2 t | t == (a, a) = -3267
      | t == (a, i) = 2744
      | t == (a, n) = -878
      | t == (h, h) = -4070
      | t == (h, m) = -1711
      | t == (h, n) = 4012
      | t == (h, o) = 3761
      | t == (i, a) = 1327
      | t == (i, h) = -1184
      | t == (i, i) = -1332
      | t == (i, k) = 1721
      | t == (i, o) = 5492
      | t == (k, i) = 3831
      | t == (k, k) = -8741
      | t == (m, h) = -3132
      | t == (m, k) = 3334
      | t == (o, o) = -2920
{-# INLINE bc2 #-}

bc3 t | t == (h, h) = 996
      | t == (h, i) = 626
      | t == (h, k) = -721
      | t == (h, n) = -1307
      | t == (h, o) = -836
      | t == (i, h) = -301
      | t == (k, k) = 2762
      | t == (m, k) = 1079
      | t == (m, m) = 4034
      | t == (o, a) = -1652
      | t == (o, h) = 266
{-# INLINE bc3 #-}

bp1 t | t == (b, b) = 295
      | t == (o, b) = 304
      | t == (o, o) = -125
      | t == (u, b) = 352
{-# INLINE bp1 #-}

bp2 t | t == (b, o) = 60
      | t == (o, o) = -1762
{-# INLINE bp2 #-}

bq1 t | t == (b, h, h) = 1150
      | t == (b, h, m) = 1521
      | t == (b, i, i) = -1158
      | t == (b, i, m) = 886
      | t == (b, m, h) = 1208
      | t == (b, n, h) = 449
      | t == (b, o, h) = -91
      | t == (b, o, o) = -2597
      | t == (o, h, i) = 451
      | t == (o, i, h) = -296
      | t == (o, k, a) = 1851
      | t == (o, k, h) = -1020
      | t == (o, k, k) = 904
      | t == (o, o, o) = 2965
{-# INLINE bq1 #-}

bq2 t | t == (b, h, h) = 118
      | t == (b, h, i) = -1159
      | t == (b, h, m) = 466
      | t == (b, i, h) = -919
      | t == (b, k, k) = -1720
      | t == (b, k, o) = 864
      | t == (o, h, h) = -1139
      | t == (o, h, m) = -181
      | t == (o, i, h) = 153
      | t == (u, h, i) = -1146
{-# INLINE bq2 #-}

bq3 t | t == (b, h, h) = -792
      | t == (b, h, i) = 2664
      | t == (b, i, i) = -299
      | t == (b, k, i) = 419
      | t == (b, m, h) = 937
      | t == (b, m, m) = 8335
      | t == (b, n, n) = 998
      | t == (b, o, h) = 775
      | t == (o, h, h) = 2174
      | t == (o, h, m) = 439
      | t == (o, i, i) = 280
      | t == (o, k, h) = 1798
      | t == (o, k, i) = -793
      | t == (o, k, o) = -2242
      | t == (o, m, h) = -2402
      | t == (o, o, o) = 11699
{-# INLINE bq3 #-}

bq4 t | t == (b, h, h) = -3895
      | t == (b, i, i) = 3761
      | t == (b, i, i) = -4654
      | t == (b, i, i) = 1348
      | t == (b, k, k) = -1806
      | t == (b, m, m) = -3385
      | t == (b, o, o) = -12396
      | t == (o, a, a) = 926
      | t == (o, h, h) = 266
      | t == (o, h, h) = -2036
      | t == (o, n, n) = -973
{-# INLINE bq4 #-}

bw1 t | t == (ord ',', ord 'と') = 660
      | t == (ord ',', ord '同') = 727
      | t == (b1, ord 'あ') = 1404
      | t == (b1, ord '同') = 542
      | t == (ord '、', ord 'と') = 660
      | t == (ord '、', ord '同') = 727
      | t == (ord '」', ord 'と') = 1682
      | t == (ord 'あ', ord 'っ') = 1505
      | t == (ord 'い', ord 'う') = 1743
      | t == (ord 'い', ord 'っ') = -2055
      | t == (ord 'い', ord 'る') = 672
      | t == (ord 'う', ord 'し') = -4817
      | t == (ord 'う', ord 'ん') = 665
      | t == (ord 'か', ord 'ら') = 3472
      | t == (ord 'が', ord 'ら') = 600
      | t == (ord 'こ', ord 'う') = -790
      | t == (ord 'こ', ord 'と') = 2083
      | t == (ord 'こ', ord 'ん') = -1262
      | t == (ord 'さ', ord 'ら') = -4143
      | t == (ord 'さ', ord 'ん') = 4573
      | t == (ord 'し', ord 'た') = 2641
      | t == (ord 'し', ord 'て') = 1104
      | t == (ord 'す', ord 'で') = -3399
      | t == (ord 'そ', ord 'こ') = 1977
      | t == (ord 'そ', ord 'れ') = -871
      | t == (ord 'た', ord 'ち') = 1122
      | t == (ord 'た', ord 'め') = 601
      | t == (ord 'っ', ord 'た') = 3463
      | t == (ord 'つ', ord 'い') = -802
      | t == (ord 'て', ord 'い') = 805
      | t == (ord 'て', ord 'き') = 1249
      | t == (ord 'で', ord 'き') = 1127
      | t == (ord 'で', ord 'す') = 3445
      | t == (ord 'で', ord 'は') = 844
      | t == (ord 'と', ord 'い') = -4915
      | t == (ord 'と', ord 'み') = 1922
      | t == (ord 'ど', ord 'こ') = 3887
      | t == (ord 'な', ord 'い') = 5713
      | t == (ord 'な', ord 'っ') = 3015
      | t == (ord 'な', ord 'ど') = 7379
      | t == (ord 'な', ord 'ん') = -1113
      | t == (ord 'に', ord 'し') = 2468
      | t == (ord 'に', ord 'は') = 1498
      | t == (ord 'に', ord 'も') = 1671
      | t == (ord 'に', ord '対') = -912
      | t == (ord 'の', ord '一') = -501
      | t == (ord 'の', ord '中') = 741
      | t == (ord 'ま', ord 'せ') = 2448
      | t == (ord 'ま', ord 'で') = 1711
      | t == (ord 'ま', ord 'ま') = 2600
      | t == (ord 'ま', ord 'る') = -2155
      | t == (ord 'や', ord 'む') = -1947
      | t == (ord 'よ', ord 'っ') = -2565
      | t == (ord 'れ', ord 'た') = 2369
      | t == (ord 'れ', ord 'で') = -913
      | t == (ord 'を', ord 'し') = 1860
      | t == (ord 'を', ord '見') = 731
      | t == (ord '亡', ord 'く') = -1886
      | t == (ord '京', ord '都') = 2558
      | t == (ord '取', ord 'り') = -2784
      | t == (ord '大', ord 'き') = -2604
      | t == (ord '大', ord '阪') = 1497
      | t == (ord '平', ord '方') = -2314
      | t == (ord '引', ord 'き') = -1336
      | t == (ord '日', ord '本') = -195
      | t == (ord '本', ord '当') = -2423
      | t == (ord '毎', ord '日') = -2113
      | t == (ord '目', ord '指') = -724
      | t == (b1, ord 'あ') = 1404
      | t == (b1, ord '同') = 542
      | t == (ord '｣', ord 'と') = 1682
{-# INLINE bw1 #-}

bw2 t | t == (ord '.', ord '.') = -11822
      | t == (ord '1', ord '1') = -669
      | t == (ord '―', ord '―') = -5730
      | t == (ord '−', ord '−') = -13175
      | t == (ord 'い', ord 'う') = -1609
      | t == (ord 'う', ord 'か') = 2490
      | t == (ord 'か', ord 'し') = -1350
      | t == (ord 'か', ord 'も') = -602
      | t == (ord 'か', ord 'ら') = -7194
      | t == (ord 'か', ord 'れ') = 4612
      | t == (ord 'が', ord 'い') = 853
      | t == (ord 'が', ord 'ら') = -3198
      | t == (ord 'き', ord 'た') = 1941
      | t == (ord 'く', ord 'な') = -1597
      | t == (ord 'こ', ord 'と') = -8392
      | t == (ord 'こ', ord 'の') = -4193
      | t == (ord 'さ', ord 'せ') = 4533
      | t == (ord 'さ', ord 'れ') = 13168
      | t == (ord 'さ', ord 'ん') = -3977
      | t == (ord 'し', ord 'い') = -1819
      | t == (ord 'し', ord 'か') = -545
      | t == (ord 'し', ord 'た') = 5078
      | t == (ord 'し', ord 'て') = 972
      | t == (ord 'し', ord 'な') = 939
      | t == (ord 'そ', ord 'の') = -3744
      | t == (ord 'た', ord 'い') = -1253
      | t == (ord 'た', ord 'た') = -662
      | t == (ord 'た', ord 'だ') = -3857
      | t == (ord 'た', ord 'ち') = -786
      | t == (ord 'た', ord 'と') = 1224
      | t == (ord 'た', ord 'は') = -939
      | t == (ord 'っ', ord 'た') = 4589
      | t == (ord 'っ', ord 'て') = 1647
      | t == (ord 'っ', ord 'と') = -2094
      | t == (ord 'て', ord 'い') = 6144
      | t == (ord 'て', ord 'き') = 3640
      | t == (ord 'て', ord 'く') = 2551
      | t == (ord 'て', ord 'は') = -3110
      | t == (ord 'て', ord 'も') = -3065
      | t == (ord 'で', ord 'い') = 2666
      | t == (ord 'で', ord 'き') = -1528
      | t == (ord 'で', ord 'し') = -3828
      | t == (ord 'で', ord 'す') = -4761
      | t == (ord 'で', ord 'も') = -4203
      | t == (ord 'と', ord 'い') = 1890
      | t == (ord 'と', ord 'こ') = -1746
      | t == (ord 'と', ord 'と') = -2279
      | t == (ord 'と', ord 'の') = 720
      | t == (ord 'と', ord 'み') = 5168
      | t == (ord 'と', ord 'も') = -3941
      | t == (ord 'な', ord 'い') = -2488
      | t == (ord 'な', ord 'が') = -1313
      | t == (ord 'な', ord 'ど') = -6509
      | t == (ord 'な', ord 'の') = 2614
      | t == (ord 'な', ord 'ん') = 3099
      | t == (ord 'に', ord 'お') = -1615
      | t == (ord 'に', ord 'し') = 2748
      | t == (ord 'に', ord 'な') = 2454
      | t == (ord 'に', ord 'よ') = -7236
      | t == (ord 'に', ord '対') = -14943
      | t == (ord 'に', ord '従') = -4688
      | t == (ord 'に', ord '関') = -11388
      | t == (ord 'の', ord 'か') = 2093
      | t == (ord 'の', ord 'で') = -7059
      | t == (ord 'の', ord 'に') = -6041
      | t == (ord 'の', ord 'の') = -6125
      | t == (ord 'は', ord 'い') = 1073
      | t == (ord 'は', ord 'が') = -1033
      | t == (ord 'は', ord 'ず') = -2532
      | t == (ord 'ば', ord 'れ') = 1813
      | t == (ord 'ま', ord 'し') = -1316
      | t == (ord 'ま', ord 'で') = -6621
      | t == (ord 'ま', ord 'れ') = 5409
      | t == (ord 'め', ord 'て') = -3153
      | t == (ord 'も', ord 'い') = 2230
      | t == (ord 'も', ord 'の') = -10713
      | t == (ord 'ら', ord 'か') = -944
      | t == (ord 'ら', ord 'し') = -1611
      | t == (ord 'ら', ord 'に') = -1897
      | t == (ord 'り', ord 'し') = 651
      | t == (ord 'り', ord 'ま') = 1620
      | t == (ord 'れ', ord 'た') = 4270
      | t == (ord 'れ', ord 'て') = 849
      | t == (ord 'れ', ord 'ば') = 4114
      | t == (ord 'ろ', ord 'う') = 6067
      | t == (ord 'わ', ord 'れ') = 7901
      | t == (ord 'を', ord '通') = -11877
      | t == (ord 'ん', ord 'だ') = 728
      | t == (ord 'ん', ord 'な') = -4115
      | t == (ord '一', ord '人') = 602
      | t == (ord '一', ord '方') = -1375
      | t == (ord '一', ord '日') = 970
      | t == (ord '一', ord '部') = -1051
      | t == (ord '上', ord 'が') = -4479
      | t == (ord '会', ord '社') = -1116
      | t == (ord '出', ord 'て') = 2163
      | t == (ord '分', ord 'の') = -7758
      | t == (ord '同', ord '党') = 970
      | t == (ord '同', ord '日') = -913
      | t == (ord '大', ord '阪') = -2471
      | t == (ord '委', ord '員') = -1250
      | t == (ord '少', ord 'な') = -1050
      | t == (ord '年', ord '度') = -8669
      | t == (ord '年', ord '間') = -1626
      | t == (ord '府', ord '県') = -2363
      | t == (ord '手', ord '権') = -1982
      | t == (ord '新', ord '聞') = -4066
      | t == (ord '日', ord '新') = -722
      | t == (ord '日', ord '本') = -7068
      | t == (ord '日', ord '米') = 3372
      | t == (ord '曜', ord '日') = -601
      | t == (ord '朝', ord '鮮') = -2355
      | t == (ord '本', ord '人') = -2697
      | t == (ord '東', ord '京') = -1543
      | t == (ord '然', ord 'と') = -1384
      | t == (ord '社', ord '会') = -1276
      | t == (ord '立', ord 'て') = -990
      | t == (ord '第', ord 'に') = -1612
      | t == (ord '米', ord '国') = -4268
      | t == (ord '１', ord '１') = -669
      | t == (ord 'ｸ', ord 'ﾞ') = 1319
{-# INLINE bw2 #-}

bw3 t | t == (ord 'あ', ord 'た') = -2194
      | t == (ord 'あ', ord 'り') = 719
      | t == (ord 'あ', ord 'る') = 3846
      | t == (ord 'い', ord '.') = -1185
      | t == (ord 'い', ord '。') = -1185
      | t == (ord 'い', ord 'い') = 5308
      | t == (ord 'い', ord 'え') = 2079
      | t == (ord 'い', ord 'く') = 3029
      | t == (ord 'い', ord 'た') = 2056
      | t == (ord 'い', ord 'っ') = 1883
      | t == (ord 'い', ord 'る') = 5600
      | t == (ord 'い', ord 'わ') = 1527
      | t == (ord 'う', ord 'ち') = 1117
      | t == (ord 'う', ord 'と') = 4798
      | t == (ord 'え', ord 'と') = 1454
      | t == (ord 'か', ord '.') = 2857
      | t == (ord 'か', ord '。') = 2857
      | t == (ord 'か', ord 'け') = -743
      | t == (ord 'か', ord 'っ') = -4098
      | t == (ord 'か', ord 'に') = -669
      | t == (ord 'か', ord 'ら') = 6520
      | t == (ord 'か', ord 'り') = -2670
      | t == (ord 'が', ord ',') = 1816
      | t == (ord 'が', ord '、') = 1816
      | t == (ord 'が', ord 'き') = -4855
      | t == (ord 'が', ord 'け') = -1127
      | t == (ord 'が', ord 'っ') = -913
      | t == (ord 'が', ord 'ら') = -4977
      | t == (ord 'が', ord 'り') = -2064
      | t == (ord 'き', ord 'た') = 1645
      | t == (ord 'け', ord 'ど') = 1374
      | t == (ord 'こ', ord 'と') = 7397
      | t == (ord 'こ', ord 'の') = 1542
      | t == (ord 'こ', ord 'ろ') = -2757
      | t == (ord 'さ', ord 'い') = -714
      | t == (ord 'さ', ord 'を') = 976
      | t == (ord 'し', ord ',') = 1557
      | t == (ord 'し', ord '、') = 1557
      | t == (ord 'し', ord 'い') = -3714
      | t == (ord 'し', ord 'た') = 3562
      | t == (ord 'し', ord 'て') = 1449
      | t == (ord 'し', ord 'な') = 2608
      | t == (ord 'し', ord 'ま') = 1200
      | t == (ord 'す', ord '.') = -1310
      | t == (ord 'す', ord '。') = -1310
      | t == (ord 'す', ord 'る') = 6521
      | t == (ord 'ず', ord ',') = 3426
      | t == (ord 'ず', ord '、') = 3426
      | t == (ord 'ず', ord 'に') = 841
      | t == (ord 'そ', ord 'う') = 428
      | t == (ord 'た', ord '.') = 8875
      | t == (ord 'た', ord '。') = 8875
      | t == (ord 'た', ord 'い') = -594
      | t == (ord 'た', ord 'の') = 812
      | t == (ord 'た', ord 'り') = -1183
      | t == (ord 'た', ord 'る') = -853
      | t == (ord 'だ', ord '.') = 4098
      | t == (ord 'だ', ord '。') = 4098
      | t == (ord 'だ', ord 'っ') = 1004
      | t == (ord 'っ', ord 'た') = -4748
      | t == (ord 'っ', ord 'て') = 300
      | t == (ord 'て', ord 'い') = 6240
      | t == (ord 'て', ord 'お') = 855
      | t == (ord 'て', ord 'も') = 302
      | t == (ord 'で', ord 'す') = 1437
      | t == (ord 'で', ord 'に') = -1482
      | t == (ord 'で', ord 'は') = 2295
      | t == (ord 'と', ord 'う') = -1387
      | t == (ord 'と', ord 'し') = 2266
      | t == (ord 'と', ord 'の') = 541
      | t == (ord 'と', ord 'も') = -3543
      | t == (ord 'ど', ord 'う') = 4664
      | t == (ord 'な', ord 'い') = 1796
      | t == (ord 'な', ord 'く') = -903
      | t == (ord 'な', ord 'ど') = 2135
      | t == (ord 'に', ord ',') = -1021
      | t == (ord 'に', ord '、') = -1021
      | t == (ord 'に', ord 'し') = 1771
      | t == (ord 'に', ord 'な') = 1906
      | t == (ord 'に', ord 'は') = 2644
      | t == (ord 'の', ord ',') = -724
      | t == (ord 'の', ord '、') = -724
      | t == (ord 'の', ord '子') = -1000
      | t == (ord 'は', ord ',') = 1337
      | t == (ord 'は', ord '、') = 1337
      | t == (ord 'べ', ord 'き') = 2181
      | t == (ord 'ま', ord 'し') = 1113
      | t == (ord 'ま', ord 'す') = 6943
      | t == (ord 'ま', ord 'っ') = -1549
      | t == (ord 'ま', ord 'で') = 6154
      | t == (ord 'ま', ord 'れ') = -793
      | t == (ord 'ら', ord 'し') = 1479
      | t == (ord 'ら', ord 'れ') = 6820
      | t == (ord 'る', ord 'る') = 3818
      | t == (ord 'れ', ord ',') = 854
      | t == (ord 'れ', ord '、') = 854
      | t == (ord 'れ', ord 'た') = 1850
      | t == (ord 'れ', ord 'て') = 1375
      | t == (ord 'れ', ord 'ば') = -3246
      | t == (ord 'れ', ord 'る') = 1091
      | t == (ord 'わ', ord 'れ') = -605
      | t == (ord 'ん', ord 'だ') = 606
      | t == (ord 'ん', ord 'で') = 798
      | t == (ord 'カ', ord '月') = 990
      | t == (ord '会', ord '議') = 860
      | t == (ord '入', ord 'り') = 1232
      | t == (ord '大', ord '会') = 2217
      | t == (ord '始', ord 'め') = 1681
      | t == (ord '市', ord ' ') = 965
      | t == (ord '新', ord '聞') = -5055
      | t == (ord '日', ord ',') = 974
      | t == (ord '日', ord '、') = 974
      | t == (ord '社', ord '会') = 2024
      | t == (ord 'ｶ', ord '月') = 990
{-# INLINE bw3 #-}

tc1 t | t == (a, a, a) = 1093
      | t == (h, h, h) = 1029
      | t == (h, h, m) = 580
      | t == (h, i, i) = 998
      | t == (h, o, h) = -390
      | t == (h, o, m) = -331
      | t == (i, h, i) = 1169
      | t == (i, o, h) = -142
      | t == (i, o, i) = -1015
      | t == (i, o, m) = 467
      | t == (m, m, h) = 187
      | t == (o, o, i) = -1832
{-# INLINE tc1 #-}

tc2 t | t == (h, h, o) = 2088
      | t == (h, i, i) = -1023
      | t == (h, m, m) = -1154
      | t == (i, h, i) = -1965
      | t == (k, k, h) = 703
      | t == (o, i, i) = -2649
{-# INLINE tc2 #-}

tc3 t | t == (a, a, a) = -294
      | t == (h, h, h) = 346
      | t == (h, h, i) = -341
      | t == (h, i, i) = -1088
      | t == (h, i, k) = 731
      | t == (h, o, h) = -1486
      | t == (i, h, h) = 128
      | t == (i, h, i) = -3041
      | t == (i, h, o) = -1935
      | t == (i, i, h) = -825
      | t == (i, i, m) = -1035
      | t == (i, o, i) = -542
      | t == (k, h, h) = -1216
      | t == (k, k, a) = 491
      | t == (k, k, h) = -1217
      | t == (k, o, k) = -1009
      | t == (m, h, h) = -2694
      | t == (m, h, m) = -457
      | t == (m, h, o) = 123
      | t == (m, m, h) = -471
      | t == (n, n, h) = -1689
      | t == (n, n, o) = 662
      | t == (o, h, o) = -3393
{-# INLINE tc3 #-}
